semgrep-core/src/tainting/Dataflow_tainting.ml

```diff
@@ -349,24 +375,50 @@ let rec check_tainted_expr env exp : Taints.t * var_env =
   in
   let check_offset env = function
     | Index e -> check env e
-    | NoOffset
-    | Dot _ ->
-        (Taints.empty, env.var_env)
+    | Dot _ -> (Taints.empty, env.var_env)
   in
   let check_subexpr exp =
     match exp.e with
-    | Fetch { base = VarSpecial (This, _); offset = Dot fld; _ } ->
+    | Fetch { base = VarSpecial (This, _); offset = Dot fld :: _; _ } ->
         (* TODO: Move this to check_tainted_instr ? *)
         let taints =
           Hashtbl.find_opt env.fun_env (str_of_name fld)
           |> Option.value ~default:PM.Set.empty
           |> PM.Set.elements |> T.taints_of_pms
         in
         (taints, env.var_env)
-    | Fetch { base; offset; _ } ->
-        let base_taints, var_env = check_base env base in
-        let offset_taints, var_env = check_offset { env with var_env } offset in
-        (Taints.union base_taints offset_taints, var_env)
+    | Fetch ({ base; offset; _ } as lval) -> (
+        let dotted_lvars = IL_lvalue_helpers.dotted_lvars_of_lval lval in
+        let x =
+          match dotted_lvars with
+          | _ :: _ as ns ->
+              let rec go = function
+                | [] -> None
+                | Some t :: _ -> Some t
+                | None :: ts -> go ts
+              in
+              ns
+              |> Common.map (fun x ->
+                     VarMap.find_opt (str_of_name x) env.var_env)
+              |> go
+          | [] -> None
+        in
+        match x with
+        | Some MarkedClean -> (Taints.empty, env.var_env)
+        | Some (Tainted _)
+        | None ->
+            let var_taints, var_env =
```

Comment:
Could we move this `let` before the `match x with` so that we don't need the `fail with "impossible"` ?